### ISSUE TITLE
Feature Request: Autoassign Asset Tag

### ISSUE THREAD

**Issue by ryanesizemore** [snipe/snipe-it#3286]
Love this project, turned out better than all of the paid services I found. I want to suggest the option for Snipe-IT to be able to automatically fill in the Asset Tag info based on a predefined scheme that the admin sets, such as "companynameXXXXX", so that the next asset automatically is set to the next number in the sequence and ready for label generation. Could also be linked to the Categories to have specific Asset Tag generations by Asset Type.

**Comment by snipe** [comment]
How would that be different than enabling auto-incrementing tags?

**Comment by ryanesizemore** [comment]
Guess it wouldn't be much different other than assigning to a category.

**Comment by therealjoshuad** [comment]
We do different prefixes per category, so for instance, assets in the computer category are PC-1000, Printers are PR-1000, so we can't use the auto increment. It leaves room for error because it's relying on looking at the last item created in a category

**Comment by faxesystem** [comment]
But wouldn't such a tag be against the 1NF? Why would you need the category or the company to be a part of the tag number?

**Comment by Go2ClassPoorYorick** [comment]
This would still be a nice feature- hopefully it can be revisited at some point.

**Comment by viclou** [comment]
+1 FD 13742

**Comment by 02ms05** [comment]
Just upgraded to new version of server (18.04 Ubuntu, PHP version 7.2.24, Snipe-IT 4.8.0)
Thing is perfect, but one thing bothers me still. 

If I have multiple categories in assets section, why can't there be auto-incrementing ID's with following prefix for each category (Pxxx for printer , Cxxx for computer, Mxxx for mobile phones etc.)?

Is there a solution for that ?

**Comment by grdzo** [comment]
We just found out about Snipe-it!! Great job!
But this feature is also of importance to us +2

- have a prefix / suffix and a starting number for each category
- autonumber based on the category and prefix/suffix information

**Comment by Mastergumble** [comment]
We also have found Snipe-IT just few days ago and it is awesome to replace our current system.

We are also looking forward to this feature, we have a lot of diferent prefixes, lap.., pc..., tlm..., srv... etc.

This would be a great feature for a lot of users.

**Comment by 02ms05** [comment]
This feature would skyrocket Snipe-IT for us, and for many others as well.

We are patiently waiting and hoping for you to implement this option, so we can add different ID's to different categories. 

So for instance: PC-1000 for PC's, LP-2000 for laptops, M-3000 for mobile phones etc., you get the idea. 

Please let me know if this will be possible in the future, we would appreciate it.

Thanks !

**Comment by newbstier** [comment]
Kind of shocked it's not a baseline function, hopefully yet.

**Comment by 02ms05** [comment]
How hard can this be to make ? There are often few major categories of assets in every company, and it can not be possible to put everything under one category (prefix). 

So please tell me is this something we can expect soon, or maybe if it's possible for us to pay you some amount to implement this? 

Please, this is the only missing puzzle for us to make the SnipeIT perfect for our corporate ITAM, and I believe bunch of other people/companies would welcome this as well.

**Comment by ghost** [comment]
is this issue resolved?

**Comment by 02ms05** [comment]
> is this issue resolved?

Still nothing to this day.

**Comment by eclark-friends** [comment]
+1 on this feature, we would love this feature.

**Comment by dpetkovoandb** [comment]
+1, it would be handy.

**Comment by Go2ClassPoorYorick** [comment]
@eclark-friends @ghost @dpetkovoandb 

Please thumb up the original thread post to raise priority for when snipe looks for features to implement.

**Comment by snipe** [comment]
I'm just not really sure how we'd implement this. An asset could potentially change companies, etc. Hell, you could change the asset model of an asset if you want - then all of your asset tags would be attributed to the wrong company/model/category/etc. 

I don't see a practical way of making this not a time bomb.

**Comment by Go2ClassPoorYorick** [comment]
I don't particularly think of those as unfair issues to encounter in that situation:

- Changing companies
If your asset numbering is based on company name, just provide a prompt to provide an asset number that's autofilled with the next number for that company

- Changing categories
Maybe my usage of the categories function is different than other people's, but I'll never have an asset change from a laptop to a desktop or to a monitor, so the pains from making it category based would be potentially self inflicted.

-Asset model
I don't see this being a practical implementation of unique asset tagging. At the point where you're tagging devices by model, may as well do categories (T450, T560, Etc)

Just my personal opinions on the matter, I'm unsure if I'm the one who's strange.
(Sorry, hit control enter because I'm used to Slack messages at this point)

**Comment by snipe** [comment]
> I don't particularly think of those as unfair issues to encounter in that situation

Not sure exactly what you mean here

**Comment by Go2ClassPoorYorick** [comment]
> > I don't particularly think of those as unfair issues to encounter in that situation
> 
> Not sure exactly what you mean here

That may be an issue compared to moving an asset now, but those are inherently questions that should be asked if you where moving an asset between companies. I understand that Snipe IT might be used in a MSP context where you might re-use a computer elsewhere, but if you've decided to implement company based asset tags, you SHOULD have to answer what to do in that situation in some way.

Edit:
Assuming that implementing of different asset tag streams is an option, and not a mandatory future.

One solution that splits the difference might be using a template for making Assets and having an option to add auto fill variables such as {{NextLaptopAsset#}} or something, but I see that being probably more of a hassle than it's worth.

Sorry, just trying to give ideas.

**Comment by snipe** [comment]
@Go2ClassPoorYorick People use companies in all kinds of different ways, not always as actual companies. But even then, I'll give you an example: 

My old ad agency was bought by a parent company. We inherited another child-company's IT department (they made us fire the consultants we were using). If we implemented something like a company-based asset tag scheme, what happens when the IT dept decides to take a laptop from one child-company and issue it to an employee of another child company? The asset tag would have to change, which means new asset labels, etc. Odds are good that they never add that new label to the laptop, and now we have mixed info, with potentially no way to find that asset ever again.

Plus what would that UI even look like? In the edit-asset page, if you select a new company, it automatically tries to find the next auto-incremented by company, I guess? But people are also asking for tags based off of category, department, etc. You can see how that gets out of hand pretty quickly, right?

**Comment by Go2ClassPoorYorick** [comment]
> If we implemented something like a company-based asset tag scheme, what happens when the IT dept decides to take a laptop from one child-company and issue it to an employee of another child company? The asset tag would have to change, which means new asset labels, etc. Odds are good that they never add that new label to the laptop, and now we have mixed info, with potentially no way to find that asset ever again.

This is a potential issue even without differing asset tags- I could easily hand out assets that are mis-tagged or not tagged at all, the fault there would be in me not receiving that laptop back to re-check it in before re-issuing. Additionally, using the original child companies asset tags (even if they had a solid system in shape) seems to me a poor idea- I would want to recount and inventory what that company already had.

I might be missing the point on this one.

> Plus what would that UI even look like? In the edit-asset page, if you select a new company, it automatically tries to find the next auto-incremented by company, I guess? But people are also asking for tags based off of category, department, etc. You can see how that gets out of hand pretty quickly, right?

I don't have a full answer to this, but in my head I see a checkmark in settings "Asset numbers by company/asset numbers by category/single stream assets"
Then, depending on the setting, the web-page detects changes in the global category setting and prompts something along the lines of 
"The new category for this asset uses a different numbering scheme. Would you like to:
-  auto number to the new category
-  manually enter a asset tag
-  leave it as is"

I'm not the most elegant at explaining the way it should look and there's nuances that I obviously can't see from here, but I imagine additional settings like "enforce asset category tagging" or something existing with it.

I definitely see this becoming more of a larger project at that point than just "numbers by category" so a quick and dirty of it could literally be on selecting a category, check the category and give it a tag from that category if one exists, else choose the next global asset if the category doesn't have a unique tag set. 

I totally get what you're saying about the options getting potentially out of hand. I imagine to prevent that you could just choose whether you wanted to develop to category based numbering or company based numbering. Both have use cases that it should be clear to why their individually beneficial. Ultimately, it's on whether you see either/both/something else as a beneficial use of time for the project.

**Comment by Go2ClassPoorYorick** [comment]
I can understand arguments against this at a fundamental level- if you look at asset tags as unique identifiers, what does it matter what the number is?

The point outside of that however, is that in actuality in a single-company setup I may be receiving support tickets from different departments with assets in them. Sure, I can pull up each asset in snipe and manually view the category, but by seeing "LAP-0004" I know I'm dealing with a laptop. For an MSP/Child company numbering scheme, I can tell at a glance where the laptop on my desk showed up at. Ultimately, it comes down to preference. Some people don't like having to pull a system up to get information like the situations described above, while others may consider it just as important to pull the asset up and get a history of it before making further decisions.

**Comment by snipe** [comment]
> if you look at asset tags as unique identifiers, what does it matter what the number is?

Because asset labels exist, and people have built integrations with bespoke internal APIs that rely on them always being consistent. Asset tags are not expected to change very much - that's why they are unique identifiers within the system.

I understand what you're asking, but I don't think you're considering all of the ways this could go very wrong - and ultimately, I'm the one that has to deal with those cases.

**Comment by Go2ClassPoorYorick** [comment]
> > if you look at asset tags as unique identifiers, what does it matter what the number is?
> 
> Because asset labels exist, and people have built integrations with bespoke internal APIs that rely on them always being consistent. Asset tags are not expected to change very much - that's why they are unique identifiers within the system.
> 
> I understand what you're asking, but I don't think you're considering all of the ways this could go very wrong - and ultimately, I'm the one that has to deal with those cases.

I was more or less alluding to that- if you ignore the human part of the asset tags they're essentially UUIDs, so it really doesn't matter what they are so long as they don't change, except that we're already allowed to manually modify asset tags so you can't in the same breath say that people rely on asset tags not changing when the paradigm inherently already allows you to change them. The difference being here a structured, automated change to the asset numbers.

And you are inherently correct- I don't have the data for how other people use the system so I'm just making a case from what I see. I can sit here and throw claims like "Why would you assume a manually editable field will always be consistent when making use of an API" but ultimately it comes down to what the majority of the user-base needs from a product- and you're in a position to speak to that.

**Comment by snipe** [comment]
I haven't given up on the idea - it's just more complicated than it seems, as most things in software are :(

**Comment by Go2ClassPoorYorick** [comment]
No worries, I'm not really trying to argue- Just giving my piece. In reality a change like this is seemingly meaningless- you overhaul an asset numbering system, how the pages save assets, what asset numbers are linked to, all to save 4 seconds while looking at an asset tag.

**Comment by snipe** [comment]
Tell that to the folks that need to print out 10k new asset labels, *and* figure out which old asset tags they belong to :(

**Comment by Go2ClassPoorYorick** [comment]
Yeah. It sucks- I'm not even dealing with asset counts of that scale, maybe 400 at most, and I'm still finding assets with tags from an earlier era. 

The thing that (pardon my insolence) confuses me here is that it feels like you're assuming that this inherently has to break work flow- if a company's just using a prefix and some digits after them, leaving this as an option (add the prefix check on category selection, move to global if null) shouldn't break any reasonable workflow at all.

**Comment by snipe** [comment]
That's not the problem really. If we add it for company, people will want it for category and department and location whatever else you can imagine. Which takes precedence if you use all of those? The amount of code we'd have to  generate - and the different cyclomatic complexity and test paths we have to maintain is a monster. :(

**Comment by Go2ClassPoorYorick** [comment]
Yep! That's certainly reason alone to leave a project like this off the docket for another time/not at all. I'd try my hand at it, but I'm trash with PHP and despite the number of responses to this don't actually have a huge amount of free time on my hands myself :rofl: 

As for precedence, I don't think there's an answer. Leaving it fully configurable changes this from a quick and dirty hackaround to a relatively tedious task. I think these are things that could ultimately be worked around if enough time was given to developing the concept, but that doesn't mean it's inherently worth that time.

**Comment by JamiePhonic** [comment]
This is a feature that i'd really like to see.

Would it be possible to have Asset Tag "Styles"? (in the same kind of way Models are defined?)
Then on the "Create Asset" screen, you'd just select the tag style you want to use for that Asset?
It would save the problem of having them coupled to a Company or Department etc
If nothing else, it would be a decent compromise for the time being.

Just my thoughts. i have no idea how this method would actually be implemented.

**Comment by Go2ClassPoorYorick** [comment]
@JamiePhonic please thumb up the original thread- it helps gain traction and visibility when features are added to projects like this.

**Comment by C-Fu** [comment]
I was pulling my hair trying to do different prefixes... then I realised that there's only one kind of prefix that we can do. Please please please have this! Just one prefix with multiple asset types seem...... odd, to say the least.

Is there an alternative method in the meantime?

**Comment by PetriAsi** [comment]
What I read above there good reasons for both cases static assets tags  and category/company/whatever based  rules.
Static tag keep identifying assets constant from systems point of view.  Category based rules are more human friendly.

What if we keep current asset tag system and implement asset **naming rules** instead?  So instead generating asset tag by rules,  system will generate asset name. That asset name could be added to printed asset labels. 

For me it would fill both needs,  treat asset tags as UUID:s and give flexibility of naming rules.

**Comment by Prathameshhankare** [comment]
I have just started to explore Snipe-IT for the organisation I am working for. I loved this project and convinced my logit team who keeps the asset records to use it and I gave a proper demo. Now, I am stuck with this just because of no category based prefix option available. They have worked with a different prefix for each category till now with the current system in place and I am finding hard to convince them to use Snipe-IT instead just because Snipe-IT doesn't have a category-based prefix option.

I need the category based prefix option to convince my team to use Snipe-IT or else I need to drop the idea altogether.

**Comment by Prathameshhankare** [comment]
Is it possible to add a prefix option in the create asset form and then add the prefix manually(of course it needs to check in the background for last used number with the prefix the user has entered) and then the form auto-assign the new incremented asset tag with the prefix?

**Comment by DomRVA** [comment]
could this not be done as part of the asset name, by category?  So the Tag is still a unique ID, but auto-increment the number in the asset name, based on category, with an optional prefix.  So, a Laptop with unique Tag ID 00054329 (or whatever) would get an asset name, based on the laptop Category, Prefix-##### where the number auto-increments.

**Comment by ecim-it** [comment]
Hello everybody! Just had following thoughts on that topic, what I think would also help a lot on site...

Yes, a custom prefix would be very helpful, since we have multiple companies to manage on one side, but on the other you have also different devices. For now - if I understand it right - the System will give the next number in line on a created asset despite the brand or company, which is very confusing.

So here me out:

I would imagine a formula-system for the auto-incrementing asset ID, which would look like follows:

1) Defining a prefix for each company (for example max 2 or 3 capitals, for "Company A" CA, for "Company B" CB). This could be accomplished and pre-defined under the details, where you create companies.

2) For each device category you could do this as well (like PR for printer or C for computer) This could be accomplished and pre-defined under the categories.

3) Other formula-options could be implemented as well, like the year, the asset was bought (last 2 numbers of the year)

4) The next number of the ID in the formula (length pre-defined - for example: 5 digits - and looked up in an table by the system)

So in the end, we could create a formula like -> company prefix+device category+year+number

This would look like CAC2100001, the next computer from Company A CAC2100001, 
but if next item is the first computer of Company B CBC00001

That way, on-site technicians and helpdesk-employees could have a glimpse of information according the ID without looking up the details.

I'm unfortunately not capable of programing, so I can just share my ideas... :-) Any supporters of that?

**Comment by PetriAsi** [comment]
@ecim-it  Your prefix generation ideas are nice and can be technically implemented. Although it's not usually wise to collect asset information to tags that should be static after creation. If there need to correct, example year device was bought, tag will change.

I'm still seeing this issue can be resolved by rules to generate device names not tags, as suggested earlier here. Rules can be complex if needed and by company  / category / etc. Then it's up to user to use device name on sticker instead of tag.

**Comment by Alaknar** [comment]
Our company has tags based on location and device type so this feature would be amazing to have.

I think the easiest way of implementing that would be to just let us set up the prefix on the asset creation screen. Say, type in the first asset's tag as `{DEVLOC}0001` so Snipe-IT could pull whatever's in the curly braces, copy it over and then increment the proceeding number.

**Comment by marcoezzy** [comment]
We need to create tags with a fix name, category, asset tag and purchase year... so I would also leave a +1 for this feature...

**Comment by QveenSi-zz** [comment]
Yes, yes, we also have different tag prefixes based on the category.
+1

**Comment by mni1000** [comment]
+1 on this, any updates?

**Comment by jwaller-ecc** [comment]
Also +1.

**Comment by achuet** [comment]
Any updates on this feature, please

**Comment by noaht8um** [comment]
+1
Use case is asset management for multiple companies as an MSP (Managed Services Provider).

**Comment by datavizzard** [comment]
I agree with most of the previous speakers here, for me one prefix per company would be enough. Then the asset number would remain the same for the internal system but each asset would be more clearly assignable. That is a real problem. This also affects companies that have several facilities or locations.

This is a real problem and prevents the implementation of Snipe It for some.

**Comment by MarcusGHub** [comment]
+1 for separate prefix per company and retain the incrementing number.

**Comment by snipe** [comment]
Once again, I will reiterate that this is much more complicated than it seems, especially based on some of the use-cases presented here. We currently store the "next increment" number in the settings table. To split this out into department, location, model, category, and all of the other things being proposed here adds a ton of complexity (and some of you want the option to switch back and forth between one or the other.)  There is nothing stopping anyone from manually entering in those location/dept/whatever specific tags when they create an asset, but when using auto-incrementing, we'd need to maintain increments for each of these different domains. And some folks want the increment to work normally (ignoring the company/dept/etc constraint) and some people don't. 

So for example, let's say this was implemented by company (since that seems the most common use case). Some folks want this:

**Case 1:**
- CompanyA-123
- CompanyB-124
- CompanyA-125
- CompanyA-126
- CompanyB-127

and other folks want 

**Case 2:**
- CompanyA-123
- CompanyB-123
- CompanyA-124
- CompanyB-124

Now read back through the use-cases above and you can see how this is a much messier thing to implement (and then test for) than it seems on the face of it. 

At the end of the day, the asset tag itself shouldn't matter. As someone else previously said, it's effectively just a UUID and doesn't (or shouldn't) have any inherent value. Case 1 is much easier to implement, but based on the feedback in this thread, there's a mix of folks who want case 1 and those who want case 2.

**Comment by MarcusGHub** [comment]
Case 2 would make more sense after reading your detailed explanation. The alternative now is that we may have to spin up separate servers for each company.

**Comment by datavizzard** [comment]
@snipe 
I would welcome the possibility to define a prefix for the tag depending on the site. The implementation with the second scenario makes sense, but I also understand the difficulty behind the decision.

But I would also be in favor of Case 2 like @MarcusGHub, which covers most of the cases I know.

Technically, case 1 makes sense, of course.

**Comment by davidebonavita** [comment]
I'll add my use case as well. 
Our company historically used tags with the format XXX0000.
We've been lucky enough to acquire two companies in the past years, each one with its load of hundreds of assets that we then imported in snipeit using their asset tags, e.g. COA0000 and COB0000.

We can't really change the asset tags for these acquired assets as they're used in several systems and most of them have a physical tag attached to them, so we need to keep using their historical format.

I've added the companies in SnipeIT, so I'm able to select them when I create a new asset. It would be great to have the asset tag prefilled with the right tag prefix (XXX, COA, COB) in the asset creation form, depending by the selected company.

Both cases suggested by @snipe in the previous comment would work for us.

**Comment by lucalehner** [comment]
I would also opt for Case 2. As an MSP, most of our customers currently use their own asset management systems, each with an auto-incremented number in their CMDB. With Case 2, it would be much easier to migrate existing assets to Snipe-IT, as each asset could retain its unique identifier. By simply adding a unique prefix for each customer, we could efficiently transfer all customer assets into Snipe-IT, maintaining consistency and organization.

Thanks @snipe for checking this feature request.

**Comment by carlashley** [comment]
I'm not sure I'd personally care if it was Case A or B per https://github.com/snipe/snipe-it/issues/3286#issuecomment-2299295142 - fundamentally all we want is something that can allow us to visually distinguish an asset tag for an asset in one company from another.

You can allocate assets to companies in Snipe-IT already, the problem is that asset tags don't have a per comapny configuration, so any asset tag generated for an asset in Company B will have the same asset tag prefix, config, etc, as an asset in Company B, _unless_ you **manually** change it.

We don't want to manually change asset tags each time we create an asset for a company because this just introduces an additional layer where human error _could_ creep in. We'd love to simply have someone select the right company and then have the asset tag update to use the right prefix and incremented asset number.

**Comment by snipe** [comment]
I'm closing this one in favor of https://github.com/grokability/snipe-it/issues/8385, which gets into a bit more of the nuance as it relates to actually implementing this.

**Comment by thanvanhung1991phidoi-png** [comment]
I have an idea like this
[https://github.com/grokability/snipe-it/issues/8385#issuecomment-4756184544](url)
